### PR TITLE
Add model def objectid support

### PR DESCRIPTION
**Pull request by superkhau**
Short term fix for ObjectId support.

Connect to strongloop-internal/scrum-loopback#625
See also https://github.com/strongloop-internal/scrum-loopback/issues/566

### CODE REVIEW

**Comment by superkhau** [comment]
@ritch Review please.

**Comment by superkhau** [comment]
@slnode test please

**Comment by superkhau** [comment]
@bajtos Review please.

File: lib/objectid.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+var ObjectId = require('mongodb').ObjectId;
```
Comment (ritch) [inline lib/objectid.js]: I don't think we should depend on `mongodb` directly in the juggler. We should contribute `ObjectID` some other way. It should work without the `mongodb` connector (eg. in the browser).

Comment (ritch) [inline lib/objectid.js]: Another option is to optionally depend on `mongodb` (or similar, eg. an `ObjectID` module). Sort of like this:

```js
```js
try {
  mongodb = require('mongodb');
} catch(e) {
  if (!isNotFound(err)) { 
    throw e;
  }
}
```

Comment (superkhau) [inline lib/objectid.js]: My initial impl had our own ObjectId definition that delegated to mongo's object id, which we can move towards in the long term update. Will add it to the notes there

For the short term I will use the try/catch you suggested.

File: lib/objectid.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+var ObjectId;
+
+try {
+  ObjectId = require('mongodb').ObjectId;
+} catch(e) {
+  if (e.code === 'MODULE_NOT_FOUND')
+    throw e
+}
+
```
Comment (superkhau) [inline lib/objectid.js]: @ritch LGTY?

**Comment by bajtos** [comment]
@superkhau please make sure the redis build failure is not related to your change

File: lib/objectid.js (hunk 2 above)
Comment (bajtos) [inline lib/objectid.js]: To be honest, I don't like adding `mongodb` dependency to juggler at all. Remember, there is no guarantee that `mongodb` module referenced by juggler will be the same module as referenced by `loopback-connector-mongodb`. In fact, this will be the most frequent case.

```
app
  node_modules
    loopback-datasource-juggler
      mongodb
    loopback-connector-mongodb
      mongodb
```

As a result, you are registering a type (constructor function) that is actually never used by the connector, because the connector has it's own local type (constructor function).

Comment (superkhau) [inline lib/objectid.js]: TBH, I don't like it either, but it was the simplest way to get the full power of an ObjectId (all it's validations, unique string generation, other functions etc) in one shot. The other option is to just not use mongodb at all and come up with our own unique implementation for each function. However, this is not a 5 point job. I figure in the long term proposal task I have for this sprint, we can decide whether we want to come up with our own implementations for each function or use source code from other projects, etc.

Comment (superkhau) [inline lib/objectid.js]: It would have an API similar to https://docs.mongodb.org/manual/reference/object-id/ basically.

Comment (superkhau) [inline lib/objectid.js]: Another thing to consider is the option to allow user-configurable ObjectId types. Like in model-config.json, we allow `{"objectid": "../path/to/your/custom/objectid"}`, or something along those lines.

**Comment by superkhau** [comment]
@bajtos Is there a way to target redis specifically? All my tests are green locally.

File: lib/objectid.js (hunk 2 above)
Comment (bajtos) [inline lib/objectid.js]: Makes me wonder if there is a way how to allow `loopback-connector-mongodb` to call `modelTypes.registerType(ObjectId)` during connector/datasource initialisation. That way we can keep the implementation in mongodb and still have ObjectID available in the app.

Of course, that would not work for isomorphic app in the browser, so it would be still a short-term fix.

Anyways, I guess you spend more time thinking about this issue than I did, and if this is the best short-term solution we have, then I can live with it.

**Comment by bajtos** [comment]
> Is there a way to target redis specifically? All my tests are green locally.

Sorry for the confusion, I was referring to `dependant loopback-connector-redis`. There are more test failures reported now. They may be unrelated to this patch, just make sure to check & verify it's true.

File: lib/objectid.js (hunk 2 above)
Comment (superkhau) [inline lib/objectid.js]: >Makes me wonder if there is a way how to allow loopback-connector-mongodb to call modelTypes.registerType(ObjectId) during connector/datasource initialisation. That way we can keep the implementation in mongodb and still have ObjectID available in the app.

I have thought about this too, but figured it would be harder to implement because we don't expose that part of the lib in DAO. The long term solution I had in mind and discussed with @ritch is an app-level registry for datatypes. During boot time, we would register all basic built in datatypes, then register any custom types. Then at run time, we can dynamically register/deregister new datatypes. Something along the lines of:

`app.datatypes.add(ObjectId);
`app.datatypes.list();
`app.datatypes.remove(ObjectId);

Or use an a K/V array for datatypes, etc. We can discuss further in the long-term proposal with regards to implementations.

**Comment by superkhau** [comment]
The redis tests are failing on its own.

**Comment by superkhau** [comment]
@bajtos Reassigning to you for review as @ritch is probably too busy to review this. Please let me know how to move forward with this.

File: lib/objectid.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+var ObjectId;
+
+try {
+  ObjectId = require('mongodb').ObjectId;
+} catch(e) {
+  if (e.code === 'MODULE_NOT_FOUND')
+    throw e
```
Comment (bajtos) [inline lib/objectid.js]: Missing semicolon at EOL, please add.

File: test/e2e/model-definition.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      Todo = ds.define('Todo', {
+        id: {type: 'ObjectId', id: true}
+      });
+    });
+
+    it('should allow 12 byte strings', function() {
+      var todo = new Todo({id: 'hello1hello1'});
+
+      todo.id.should.be.an.instanceof(ObjectId);
+      todo.id.should.be.an.instanceof(MongoObjId);
```
Comment (bajtos) [inline test/e2e/model-definition.js]: The fact that juggler's `ObjectID` type is implemented via mongodb's type is IMO an implementation detail that should not be asserted in unit-tests. If we want to check that `ObjectID` behaves like mongodb's `ObjectID`, then I'd much rather add few more tests to test that behaviour.

Thoughts?

**Comment by bajtos** [comment]
@superkhauplease add few tests to the test-suite run by all connectors (see [test/common.batch.js](https://github.com/strongloop/loopback-datasource-juggler/blob/4ce02cbcc13a584390349293cfc35fb1c120b1fc/test/common.batch.js)) to ensure this new ObjectID type is correctly persisted and loaded when using e.g. relational (SQL) databases.

File: test/e2e/model-definition.js (hunk 4 above)
Comment (bajtos) [inline test/e2e/model-definition.js]: Also the tests in this file duplicates the tests in `test/objectid.test.js`. What's the point of that?

I am proposing to simplify tests in this file to a bare minimum needed in order to verify that `type: 'ObjectId'` is correctly coerced into an instance of `require('../../lib/objectid')`. Everything else should be tested in `test/objectid.test.js`.

File: lib/types.js [hunk 5]
```diff
@@ -62,6 +63,7 @@ module.exports = function (modelTypes) {
   modelTypes.registerType(Array);
   modelTypes.registerType(GeoPoint);
   modelTypes.registerType(Object);
+  modelTypes.registerType(ObjectId);
```
Comment (bajtos) [inline lib/types.js]: Is there any way how to register `ObjectID` as an alias for `ObjectId`? Or is it automatically handled?

File: test/e2e/model-definition.js (hunk 4 above)
Comment (superkhau) [inline test/e2e/model-definition.js]: >The fact that juggler's ObjectID type is implemented via mongodb's type is IMO an implementation detail that should not be asserted in unit-tests. If we want to check that ObjectID behaves like mongodb's ObjectID, then I'd much rather add few more tests to test that behaviour.

I can live with this. We shouldn't test MongoDB's impl at all since it is an impl detail of our custom ObjectId type. Will remove tests that relate to MongoObjId.

Comment (superkhau) [inline test/e2e/model-definition.js]: >Also the tests in this file duplicates the tests in test/objectid.test.js. What's the point of that?

They are testing from outside in vs inside out (ie. functional vs unit). They look like dupes, but should be tested independently. The e2e tests use things like `new Todo({type: 'ObjectId', ...})` vs `new ObjectId()`. IMO, LoopBack mixes unit/functional tests together way too much and shouldn't be. The different aspects should be tested independently and in isolation.

File: lib/types.js (hunk 5 above)
Comment (superkhau) [inline lib/types.js]: It is already registered as an alias in the ObjectId class definition, so it should work properly. I added an e2e test to prove it.

File: test/e2e/model-definition.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    });
+  });
+
+  context('ObjectID (alias) definition', function() {
+    var Todo;
+
+    before(function setup() {
+      var ds = new DataSource();
+      Todo = ds.define('Todo', {
+        id: {type: 'ObjectID', id: true}
```
Comment (superkhau) [inline test/e2e/model-definition.js]: Test added here.

**Comment by superkhau** [comment]
@bajtos Updated with the comments from your review. Can you PTAL again?

File: lib/objectid.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+var ObjectId;
+
+try {
+  ObjectId = require('mongodb').ObjectId;
+} catch(e) {
+  if (e.code === 'MODULE_NOT_FOUND')
+    throw e;
```
Comment (bajtos) [inline lib/objectid.js]: @superkhau AFAICT, this silently swallows any `mongodb` parse errors, which does not look like a good thing to me. Could you please explain what is the rationale for this, why we need to give `require('mongodb').ObjectId` errors a special treatment?

**Comment by bajtos** [comment]
@superkhau I am missing unit-tests to be run by connectors, per my earlier comment:

> please add few tests to the test-suite run by all connectors (see [test/common.batch.js](https://github.com/strongloop/loopback-datasource-juggler/blob/4ce02cbcc13a584390349293cfc35fb1c120b1fc/test/common.batch.js)) to ensure this new ObjectID type is correctly persisted and loaded when using e.g. relational (SQL) databases.

File: lib/objectid.js (hunk 7 above)
Comment (superkhau) [inline lib/objectid.js]: This was suggested by @ritch in a [previous comment in this issues history](https://github.com/strongloop/loopback-datasource-juggler/pull/778#discussion_r46341514). I think it was for uses cases where mongodb is not available (ie. the browser).

**Comment by superkhau** [comment]
@slnode test please

File: test/datatype.test.js [hunk 8]
```diff
@@ -3,7 +3,30 @@ var should = require('./init.js');
 
 var db, Model;
 
-describe('datatypes', function () {
```
Comment (superkhau) [inline test/datatype.test.js]: @bajtos Added data type tests here.

**Comment by superkhau** [comment]
@bajtos Added data type tests here: https://github.com/strongloop/loopback-datasource-juggler/pull/778/files#r47993341

File: lib/objectid.js (hunk 7 above)
Comment (bajtos) [inline lib/objectid.js]: Fair enough. Please add a debug log to let the user know `ObjectId` will be undefined because `mongodb` was not found.

```js
debug('Cannot load mongodb, ObjectId will be undefined. %s', e.stack);
```

**Comment by bajtos** [comment]
One more comment: https://github.com/strongloop/loopback-datasource-juggler/pull/778#discussion_r48249729

@superkhau after you address ^^^, then please rebase the patch on top of the current master, clean up the commit history and merge this without waiting for me. IIUC this is a backwards-compatible change targeted for LoopBack LTS, in which case you need to back-port it to `2.x` too.

**Comment by bajtos** [comment]
`2.x` refers to this: https://github.com/strongloop/loopback-datasource-juggler/tree/2.x

**Comment by mrbatista** [comment]
@superkhau when will this patch be merged?

Thanks.

**Comment by superkhau** [comment]
@mrbatista Soon I believe, I need to write a few more tests to confirm the data with relational connectors, but after that it's all good.

**Comment by bajtos** [comment]
@superkhau I am not sure what's the status of this work. If we decide to eventually land this, then I would like you to try an alternative approach: get the implementation of `ObjectID` directly from [bson/lib/bson/objectid.js](https://github.com/mongodb/js-bson/blob/66bbd418c711d8b5a7f3510933fe0840a3e12b53/lib/bson/objectid.js). A believe the `mongodb` driver we use does basically the same, so it should be safe to bypass `mongodb` and load the definition directly. If it works, then this solution will reduce the footprint of juggler in applications that don't use MongodB.

**Comment by superkhau** [comment]
@bajtos This has been stale for a long time and is probably not a good solution now. Let's do what you propose instead.

**Comment by bajtos** [comment]
ok, I'm closing this PR then.
